src/mongo/db/pipeline/expression.cpp

```diff
@@ -2716,31 +2731,101 @@ Value ExpressionIndexOfArray::evaluate(const Document& root) const {
+            return Value(static_cast<int>(i));
+        }
+    }
 
-    Value searchItem = vpOperand[1]->evaluate(root);
+    return Value(-1);
+}
 
-    size_t startIndex = 0;
-    if (vpOperand.size() > 2) {
-        Value startIndexArg = vpOperand[2]->evaluate(root);
+vector<Value> ExpressionIndexOfArray::evaluateAndValidateArguments(const Document& root,
+                                                const ExpressionVector& operands,
+                                                size_t arrayLength) const {
+    std::vector<Value> deps;
+    deps.push_back(operands[1]->evaluate(root));
+
+    int startIndex = 0;
+    if (operands.size() > 2) {
+        Value startIndexArg = operands[2]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(startIndexArg, getOpName(), "starting index");
-        startIndex = static_cast<size_t>(startIndexArg.coerceToInt());
+        startIndex = static_cast<int>(startIndexArg.coerceToInt());
     }
+    deps.push_back(Value(startIndex));
 
-    size_t endIndex = array.size();
-    if (vpOperand.size() > 3) {
-        Value endIndexArg = vpOperand[3]->evaluate(root);
+    int endIndex = arrayLength;
+    if (operands.size() > 3) {
+        Value endIndexArg = operands[3]->evaluate(root);
         uassertIfNotIntegralAndNonNegative(endIndexArg, getOpName(), "ending index");
         // Don't let 'endIndex' exceed the length of the array.
-        endIndex = std::min(array.size(), static_cast<size_t>(endIndexArg.coerceToInt()));
+        endIndex = std::min(arrayLength, static_cast<size_t>(endIndexArg.coerceToInt()));
+    }
+    deps.push_back(Value(endIndex));
+    return deps;
+}
+/**
+ * This class handles the case where IndexOfArray is given an ExpressionConstant
+ * instead of using a vector and searching through it we can use a unordered_map
+ * for O(1) lookup time.
+ */
+class ExpressionIndexOfArray::Optimized : public ExpressionIndexOfArray {
+public:
+    Optimized(const boost::intrusive_ptr<ExpressionContext>& expCtx,
+              const ValueUnorderedMap<int>& indexMap,
+              const ExpressionVector& operands)
+        : ExpressionIndexOfArray(expCtx), _indexMap(std::move(indexMap)) {
+        vpOperand = operands;
     }
 
-    for (size_t i = startIndex; i < endIndex; i++) {
-        if (getExpressionContext()->getValueComparator().evaluate(array[i] == searchItem)) {
-            return Value(static_cast<int>(i));
+    virtual Value evaluate(const Document& root) const {
+        std::vector<Value> operands = evaluateAndValidateArguments(root, vpOperand, _indexMap.size());
+        auto index = _indexMap.find(operands[0]);
+        auto comparator = getExpressionContext()->getValueComparator();
+        Value vIndex = Value(index->second);
+        if (index != _indexMap.end() && comparator.evaluate(Value(index->second) >= operands[1]) &&
+            comparator.evaluate(Value(index->second) < operands[2])) {
```

Comment:
Oh this is pretty awkward. You have to keep using `comparator.evaluate(...)` because `operands[1]` and `operands[2]` are of type `Value`. In reality, we know they will be of type `int`, right? Or even `size_t`?

Can we change the signature of `evaluateAndValidateArguments` to return a struct, something like
```
struct Arguments {
    Arguments(Value targetOfSearch, int startIndex, boost::optional<int> endIndex) : 
        targetOfSearch(targetOfSearch), startIndex(startIndex), endIndex(endIndex) {}

    Value targetOfSearch;
    int startIndex;
    boost::optional<int> endIndex;
};
```

It might be more complicated in the parsing code, but I think it would make this much easier to read.

Reply:
Also, how do you know that operands[2] is valid? Couldn't the user have left off the end index?